tools/pyboard.py: Remove Python 3.9 dependency.

jimmo opened the pull request:
Changes the use of BooleanOptionalAction to a way that works on previous Python versions.

Fixes #7700 and https://github.com/micropython/micropython/pull/7459#issuecomment-876421056

dpgeorge commented:
Hmm, it looks like there's a problem with the mutually-exclusive arg group that has crept in here:
- 178198a01df51b5f4c5ef9f38ab2fb8f6269d5f4 accidentally added the new `--no-exclusive` option to this group
- 41adf178309759d5965c15972f04987a2635314c then also added the new `--no-soft-reset` option to the same group
- this PR adds a further option to this group

The group should only be used by `--follow` and `--no-follow` because it only makes sense to supply one of those options not both (see original commit for this 83afd48ad93e162eec65d34c2dfe266c995fbafd).

Steps that need to be taken:
- this PR should create a new group for the `--soft-reset` and `--no-soft-reset` options
- another commit (in this PR or another one) should move `--no-exclusive` out to the top-level `cmd_parser` object (or create a new mut-ex group which also adds `--exclusive` for consistency)

jimmo committed:
tools/pyboard.py: The --no-exclusive flag shouldn't be mutexed.

See 178198a.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/pyboard.py: The --soft-reset shouldn't be mutexed.

See 41adf17, where it was added to the mutex group accidentally.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/pyboard.py: Make --no-follow use --follow.

You can set one or the other (or neither) but not both.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/pyboard.py: Add --exclusive to match --no-exclusive.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/pyboard.py: Makes --no-soft-reset consistent with other args.

Makes it work like --no-follow and --no-exclusive using a mutex group and
dest. Although the current implementation with BooleanOptionAction is
neater, it requires Python 3.9.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/pyboard.py: Clarify defaults for soft-reset/exclusive.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
I've done both of those steps. Probably went a bit overboard with the commits, feel free to squash as necessary.

dpgeorge commented:
This was squashed to 4 commits and merged in 2a290bbfe18929fdd00e24d937d9dfc6530eb874 through 145fedef8d74a504c4b7c000e6c305602d07feff

Thank you!
